payments_redsys/__init__.py

```diff
@@ -1,259 +1,292 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with django-payments-redsys.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+import base64
+import hashlib
+import hmac
+import json
+import logging
+import re
+
+import pyDes
+import requests
+from django import forms
+from django.http import HttpResponseRedirect
+from payments import PaymentError
+from payments.core import BasicProvider, get_base_url, urljoin
+from payments.forms import PaymentForm
+
+logger = logging.getLogger(__name__)
+
+# https://en.wikipedia.org/wiki/ISO_4217
+ISO_CURRENCY_LOOKUP = {
+    "EUR": "978",
+    "GBP": "826",
+    "USD": "840",
+}
+
+
+def compute_signature(salt: str, payload: bytes, key: str):
+    """
+    For Redsys:
+        salt = order number (Ds_Order or Ds_Merchant_Order)
+        payload = Ds_MerchantParameters
+        key = shared secret (aka key) from the Redsys Administration Module
+              (Merchant Data Query option in the "See Key" section)
+    """
+    bkey = base64.b64decode(key)
+    des3 = pyDes.triple_des(
+        bkey, mode=pyDes.CBC, IV="\0" * 8, pad="\0", padmode=pyDes.PAD_NORMAL
+    )
+    pepper = des3.encrypt(str(salt))
+    payload_hash = hmac.new(pepper, payload, hashlib.sha256).digest()
+    return base64.b64encode(payload_hash)
+
+
+def compare_signatures(sig1, sig2):
+    alphanumeric_characters = re.compile("[^a-zA-Z0-9]")
+    sig1safe = re.sub(alphanumeric_characters, "", sig1)
+    sig2safe = re.sub(alphanumeric_characters, "", sig2)
+    return sig1safe == sig2safe
+
+
+class RedsysResponseForm(forms.Form):
+    Ds_SignatureVersion = forms.CharField(max_length=256)
+    Ds_Signature = forms.CharField(max_length=256)
+    Ds_MerchantParameters = forms.CharField(max_length=2048)
+
+
+REDSYS_ENVIRONMENTS = {
+    "real": "https://sis.redsys.es",
+    "test": "https://sis-t.redsys.es:25443",
+}
+
+
+class RedsysProvider(BasicProvider):
+    def __init__(self, *args, **kwargs):
+        self.language = kwargs.pop("language")
+        self.merchant_code = kwargs.pop("merchant_code")
+        self.terminal = kwargs.pop("terminal")
+        self.shared_secret = kwargs.pop("shared_secret")
+        self.currency = kwargs.pop("currency", "978")
+        self.direct_payment = str(kwargs.pop("direct_payment", False)).upper()
+        self.endpoint = REDSYS_ENVIRONMENTS[kwargs.pop("environment", "test")]
+        self.order_number_prefix = kwargs.pop("order_number_prefix", "0000")
+        self.order_number_min_length = kwargs.pop("order_number_min_length", 0)
+        self.signature_version = kwargs.pop("signature_version", "HMAC_SHA256_V1")
+        super(RedsysProvider, self).__init__(*args, **kwargs)
+
+    def get_form(self, payment, data=None):
```

Comment:
I've reordered the methods so that the public ones called by django payments are on top. I realise looking at the diff won't be very useful, so you may want to just take a look at the full new file.